[4.0] Fixing removeOrphanNodes() in Finder Taxonomy

Hackwar opened the pull request:
When an element is indexed, Finder then executes an optimisation step. In that step it removes all empty taxonomies and since the current code simply looks for direct mappings between indexed items and taxonomies and not also between items and childs of taxonomies, it is deleting empty taxonomies like sample-data-articles, which results in failures afterwards.

This code here should in theory fix that, but unfortunately I get a strange error, where $this in DatabaseDriver all of a sudden is null. Please test anyway.

Hackwar committed:
Fixing removeOrphanNodes() in Finder Taxonomy

alikon commented:
test was made  on postgresql
i've applyied the patch on a fresh install  and then i've runned the Testing Sample Data module/plugin
`2019-09-23 18:46:13.954 CEST [4912] postgres@joomla ERROR:  invalid input syntax for integer: "" at character 57
2019-09-23 18:46:13.954 CEST [4912] postgres@joomla STATEMENT:  SELECT *
	FROM "my4_finder_taxonomy"
	WHERE "parent_id" = '' AND "title" = 'Scenery' AND "language" = ''`

hoping it helps to discover the root of the bug

wilsonge committed:
Merge branch '4.0-dev' into patch-23

SharkyKZ commented on administrator/components/com_finder/Indexer/Taxonomy.php:
```diff
@@ -358,27 +358,33 @@ public static function removeMaps($linkId)
-			->where($db->quoteName('id') . ' IN (' . $subquery . ')');
+		do
+		{
+			$nodes = $db->loadColumn();
```
Move `$db->setQuery($query)` before this line to fix `Call to a member function execute() on null`. Query needs to be re-set when loading results but not when running `execute()`. Not sure if this is intended or something that needs to be fixed in Database package.

Hackwar replied:
It works in 3.x, so looks like an issue with the database package.

Hackwar replied:
hmm, not really sure. Idk. Going to fix only this PR.

Hackwar replied:
Done

Hackwar committed:
Moving setQuery()

Hackwar commented:
@alikon Thank you for this, I will definitely fix that, however that bug is unrelated to this PR/bug. Could you open a new issue for that and mention me there?

Hackwar commented:
Seems as if this currently fixes the bug, but the number of articles results in a maxexecution time. Maybe someone could split up the article step further?

Hackwar committed:
Merge branch '4.0-dev' into patch-23

wilsonge committed:
Merge branch '4.0-dev' into patch-23

infograf768 commented:
I still get an error after this patch on current 4.0-dev, but this time the error is readable

N6REJ commented:
> Seems as if this currently fixes the bug, but the number of articles results in a maxexecution time. Maybe someone could split up the article step further?

with only 60 articles?... wow

N6REJ commented:
I have tested this item :red_circle: unsuccessfully on 4bc43b811c9eda45ba856516fcc8574a1e14b44a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26392">issues.joomla.org/tracker/joomla-cms/26392</a>.</sub>

wilsonge commented:
I'm with @N6REJ no issues with execution times etc. But it doesn't give any change in the error message I'm receiving either :( My taxonomy table doesn't seem any different either - so I can only really conclude that this PR hasn't had any effect on the issue

rdeutz committed:
Merge branch '4.0-dev' into patch-23

Hackwar committed:
Merge branch '4.0-dev' into patch-23

Hackwar committed:
Merge branch '4.0-dev' into patch-23

Hackwar committed:
Merge branch '4.0-dev' into patch-23

wilsonge commented:
@Hackwar what's your plans here? Given this has been tested to not work as it stands even when the query does run

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into patch-23

Hackwar commented:
I will investigate this in the coming week and provide a fix. To be honest, I'm a little bit stumped because this is the fix that my customer got and it works there. I will have to search what I missed there. The source of the problem however is right.

wilsonge commented:
Thanks :) I can completely believe this is a decent part of the fix :)

Hackwar committed:
Fixing code error

Hackwar committed:
Merge branch '4.0-dev' into patch-23

Hackwar committed:
Don't delete taxonomies every time
Recreate the Article model when creating testing sampledata

Hackwar committed:
Merge branch 'patch-23' of github.com:Hackwar/joomla-cms into patch-23

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into patch-23

Hackwar commented:
So I found the issue, but I'm unsure how to fix it... First of all, there was a stupid little coding error, which I fixed above. But here comes the big issue:

The indexer calls the Taxonomy class to store taxonomy entries. In the store method, the code caches the used taxonomies to quickly retrieve the IDs. So now the model fires "onContentAfterSave" and the indexer runs, deletes the previous entry and indexes the new data. So far so "good". (Not exactly, but that is how it works right now...)

At some point we added onContentStateChange in a way, that it is executed when the testing sampledata plugin adds its articles. In that event, finder deletes the previous entry and reindexes it, to just then fire onContentAfterSave, deleting the entry again and doing all that indexing AGAIN, but now, since the taxonomies are identical, it runs into the issue of the cache. The previous delete also removed the then empty taxonomies, but they are still cached and thus the next indexing returns the wrong IDs and we have this nice issue here...

This also brings up the issue, that the filters, which you can create, are not going to work when the taxonomies are recreated due to such an edit. So we have to fix that one, too.

And while I'm writing all of this, I've come up with a solution. ;-) First of all, the removal of the empty taxonomies is delayed until after the reindexing is done and then the article model is recreated every time in the testing sampledata plugin to prevent it from thinking it changed state. (I did not check why that is in the model. Maybe there is a bug? Maybe we need a reset() method for the model to prevent situations like this?)

Looking at the other sampledata plugins, we need to refactor those to work similarly to the testing plugin, since those partially just insert directly into the DB.

Hackwar committed:
Merge branch '4.0-dev' into patch-23

Hackwar committed:
Merge branch '4.0-dev' into patch-23

Hackwar commented:
Could those who tested thios prior, test this again?

richard67 commented:
Testing instructions would be helpful.

Hackwar commented:
- Install a current 4.0 and install the testing sampledata.
- Notice that it errors out
- Apply this patch, reinstall the installation
- Notice that it does not fail anymore.

richard67 commented:
Normally we have testing instructions in the PR's description and not in some comment.

richard67 commented:
**Important hint for testers:** In addition to this PR, PR #27008 has to be applied, too, to make testing sample data installation work.

richard67 commented:
I have tested this item :white_check_mark: successfully on 546718232d42cc002be5e8e71f9750f18935a878<br /><br />Important note for other testers: It needs PR #27008 to be applied, too, in addition to this PR here.

Test as follows:

1. On a clean 4.0-dev installation without any sample data, switch error reporting to "Maximum" or "Development" and apply PR #27008 **but not this PR here**.

2. Install testing sample data.

Result: Testing sample data installation fails at step 4 with an error similar to the one shown in this comment: [https://github.com/joomla/joomla-cms/pull/26392#issuecomment-534644501](https://github.com/joomla/joomla-cms/pull/26392#issuecomment-534644501 "https://github.com/joomla/joomla-cms/pull/26392#issuecomment-534644501").

3. Again on a clean 4.0-dev installation without any sample data, switch error reporting to "Maximum" or "Development" and apply PR #27008 **and this PR here**.

4. Install testing sample data.

Result: Testing sample data installation finishs with success.

@Hackwar Please add my instructions to a section "Testing Instructions" in the dexription of this PR so other testers easily find it. Otherwise we might get bad tests, e.g. because people don't apply PR #27008 , too.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26392">issues.joomla.org/tracker/joomla-cms/26392</a>.</sub>

richard67 commented:
@alikon @N6REJ If you want to test this PR, please use the testing instructions I gave in my test result here: [https://github.com/joomla/joomla-cms/pull/26392#issuecomment-550491854](https://github.com/joomla/joomla-cms/pull/26392#issuecomment-550491854). Reason is that it needs PR #27008 to be applied, too, in addition to this one, and so it also needs to test that PR #27008 alone does not solve the issue with testing sample data.

Quy commented:
Also installed #27008. In the PHP error log:

> PHP Fatal error:  Maximum execution time of 30 seconds exceeded in C:\xampp\htdocs\joomla-cms-4.0-dev\libraries\vendor\joomla\database\src\Mysqli\MysqliStatement.php on line 443

![26392](https://user-images.githubusercontent.com/368084/68346390-da962480-00a8-11ea-8310-8c089dadab83.jpg)

richard67 commented:
@Quy Strange, my max_execution_time is also 30 seconds and I did not have this problem. I had a clean installation of current 4.0-dev without any other sample data or additional stuff.

Quy commented:
Local clean install with no other sample data on Windows 10 using XAMPP. No multilingual site.

richard67 commented:
I guess we both have MySQL 5.7? Or do you already use 8? If 5.7, then the only difference is the OS, it seems. Or maybe it is a random thing and it happens to me, too, when I test again? I will check tomorrow.

wilsonge commented on plugins/sampledata/testing/testing.php:
```diff
@@ -4624,12 +4624,13 @@ private function addArticles(array $articles)
 		foreach ($articles as $i => $article)
 		{
+			/** @var \Joomla\Component\Content\Administrator\Model\ArticleModel $model */
+			$model = $mvcFactory->createModel('Article', 'Administrator', ['ignore_request' => true]);
```
Why a fresh model per article?

Hackwar replied:
Because otherwise the model thinks it is updating the state of the articles and fires both the save event and the updateState event. That results in the indexing being run twice.

Quy commented:
10.4.6-MariaDB

richard67 commented:
Then it would be interesting to know if someone else with same MariaDB version has the problem, too.

wilsonge committed:
Merge branch '4.0-dev' into patch-23

richard67 commented:
**Hint for other testers:** PR #27008 is merged now, and this PR has been updated by last changes in 4.0-dev, so there is no need anymore to apply PR #27008 for testing.

wilsonge commented:
So I'm running 10.3.16-MariaDB and this patch worked for me BUT that step that you timed out on took 18 seconds on my mid-range MacBook. I can totally believe on some setups it's going to exceed 30. We'll need to spend some time figuring out exactly what's taking so long. However I think this fixes the actual error on sample data install so I'm merging this for now. @Quy could you raise an issue for this please so we can investigate the reason it's taking so long on that step to create all the articles. Thanks!

alikon commented:
if batch inserting few articles on smart search like the sample plugin do,  took 18 seconds ... then something is really bad/wrong

wilsonge commented:
There's 68 articles in the test data set plus all the associated permission to calculate and insert into the asset table. I mean it's going to take a while. But yeah 18 seconds seems too long.
